core/primitives/src/validator_mandates.rs

```diff
@@ -81,56 +84,110 @@ impl ValidatorMandates {
+        for i in 0..validators.len() {
+            let partial_weight = validators[i].partial_mandate_weight(config.stake_per_mandate);
+            if partial_weight > 0 {
+                partials.push((i as ValidatorId, partial_weight));
+            }
+        }
+
+        Self { config, mandates, partials }
     }
 
     /// Returns a validator assignment obtained by shuffling mandates.
     ///
     /// It clones mandates since [`ValidatorMandates`] is supposed to be valid for an epoch, while a
     /// new assignment is calculated at every height.
-    ///
-    /// # Interpretation of the return value
-    ///
-    /// The returned vector contains one map per shard, with the position in the vector
-    /// corresponding to `shard_id` in `0..num_shards`.
-    ///
-    /// Each `HashMap` maps `ValidatorId`s to the number of mandates they have in the corresponding
-    /// shards. A validator whose id is not in a map has not been assigned to the shard.
-    ///
-    /// ## Example
-    ///
-    /// Let `res` be the return value of this function, then `res[0][1]` maps to the number of
-    /// mandates validator with `ValidatorId` 1 holds in shard with id 0.
-    pub fn sample<R>(&self, rng: &mut R) -> Vec<HashMap<ValidatorId, u16>>
+    pub fn sample<R>(&self, rng: &mut R) -> ValidatorMandatesAssignment
     where
         R: Rng + ?Sized,
     {
-        let shuffled_mandates = self.shuffled(rng);
+        let shuffled_mandates = self.shuffled_mandates(rng);
+        let shuffled_partials = self.shuffled_partials(rng);
 
-        // Assign shuffled seat at position `i` to the shard with id `i % num_shards`.
-        let mut assignments_per_shard = Vec::with_capacity(self.config.num_shards);
+        // TODO(#10014) shuffle shard ids to avoid a bias towards smaller shard ids
+        let mut mandates_per_shard = Vec::with_capacity(self.config.num_shards);
         for shard_id in 0..self.config.num_shards {
-            let mut assignments = HashMap::new();
+            let mut assignments: HashMap<ValidatorId, AssignmentWeight> = HashMap::new();
+
+            // Assign shuffled mandates at position `i` to the shard with id `i % num_shards`.
             for idx in (shard_id..shuffled_mandates.len()).step_by(self.config.num_shards) {
                 let id = shuffled_mandates[idx];
-                assignments.entry(id).and_modify(|counter| *counter += 1).or_insert(1);
+                assignments
+                    .entry(id)
+                    .and_modify(|assignment_weight| {
+                        assignment_weight.num_mandates += 1;
+                    })
+                    .or_insert(AssignmentWeight::new(1, 0));
             }
-            assignments_per_shard.push(assignments)
+
+            // Assign shuffled partials at position `i` to the shard with id `i % num_shards`.
+            for idx in (shard_id..shuffled_partials.len()).step_by(self.config.num_shards) {
```

Comment:
I don't understand why the range starts at `shard_id`. Both in this and the previous loop.

`self.partials` is a vector of validator ids. `shard_id` and `validator_id` are both numbers but they are completely different domains.

Reply:
Here we want to distribute the elements of `shuffled_mandates` and `shuffled_partials` into `num_shards` buckets. The outer loop is over `(0..num_shards)`, so in the inner loop we fill the bucket for a particular `shard_id`. This is done by starting at `shard_id` and then taking every `num_shards`-th element, which is why we use `step_by(self.config.num_shards)`.

In a8422c0 I’ve updated and added comments.

Does that help?

Reply:
It's an alternative way to do this rather than iterating from 0 to N and at each step modding by k. Instead we use a nested loop to loop through the mod, and then iterating through [mod, k+mod, 2k+mod, ...].